packages/replicache/src/kv/sqlite-store.ts

```diff
@@ -0,0 +1,340 @@
+  ) {
+    this.#name = name;
+    this.#dbm = dbm;
+    this.#rwLock = rwLocks.get(name) ?? new RWLock();
+    rwLocks.set(name, this.#rwLock);
+
+    this.#db = this.#dbm.open(name, opts);
+
+    this.#preparedStatements = getPreparedStatementsForSQLiteDatabase(this.#db);
+  }
+
+  async read(): Promise<Read> {
+    const release = await this.#rwLock.read();
+    try {
+      const read = new SQLiteStoreRead(this.#preparedStatements, release);
+
+      return Promise.resolve(read);
+    } catch (e) {
+      return Promise.reject(e);
+    }
+  }
+
+  async write(): Promise<Write> {
+    const release = await this.#rwLock.write();
+    try {
+      const write = new SQLiteStoreWrite(this.#preparedStatements, release);
+
+      return Promise.resolve(write);
+    } catch (e) {
+      return Promise.reject(e);
+    }
+  }
+
+  close(): Promise<void> {
+    for (const stmt of Object.values(this.#preparedStatements)) {
+      stmt.finalize();
+    }
+
+    this.#dbm.close(this.#name);
+    this.#closed = true;
+
+    return promiseVoid;
+  }
+
+  get closed(): boolean {
+    return this.#closed;
+  }
+}
+
+export class SQLiteStoreRead implements Read {
+  #closed = false;
+  protected _preparedStatements: SQLitePreparedStatements;
+  readonly #release: () => void;
+
+  constructor(
+    preparedStatements: SQLitePreparedStatements,
+    release: () => void,
+  ) {
+    this._preparedStatements = preparedStatements;
+    this.#release = release;
+  }
+
+  has(key: string): Promise<boolean> {
+    const unsafeValue = this.#getSql(key);
+    return Promise.resolve(unsafeValue !== undefined);
+  }
+
+  get(key: string): Promise<ReadonlyJSONValue | undefined> {
+    const unsafeValue = this.#getSql(key);
+    if (unsafeValue === undefined) return promiseUndefined;
+    const parsedValue = JSON.parse(unsafeValue) as ReadonlyJSONValue;
+    const frozenValue = deepFreeze(parsedValue);
+    return Promise.resolve(frozenValue);
+  }
+
+  release(): void {
+    this.#closed = true;
+    this.#release();
+  }
+
+  get closed(): boolean {
+    return this.#closed;
+  }
+
+  #getSql(key: string): string | undefined {
+    const rows = this._preparedStatements.get.all<{value: string}>([key]);
+
+    if (rows.length === 0) return undefined;
+
+    return rows?.[0]?.value;
+  }
+}
+
+export class SQLiteStoreWrite extends WriteImplBase implements Write {
+  readonly #preparedStatements: SQLitePreparedStatements;
+
+  constructor(
+    preparedStatements: SQLitePreparedStatements,
+    release: () => void,
+  ) {
+    super(new SQLiteStoreRead(preparedStatements, release));
+    this.#preparedStatements = preparedStatements;
+  }
+
+  commit(): Promise<void> {
+    if (this._pending.size === 0) return promiseVoid;
+
+    // we bundle writes together and run the commit in a synchronous block
+    const next = new Promise<void>(resolve => {
+      // BEGIN IMMEDIATE grabs an RESERVED lock
+      this.#preparedStatements.beginImmediate.run();
```

Comment:
why do we only begin the sqlite transaction on commit?

Yes, the read-write locks should be ensuring proper concurrency control (so long as no other _process_ is interacting with the db) but, for performance, it seems that we would want to scope the transaction around all operations in `withWrite` and `withRead` as each `read` operation will open an implicit transaction if one is not already open.

Reply:
Yeah this is good point. I will continue to look into scoping the transaction higher - originally there were issues with sharing a single SQLite connection across the StoreProvider. The async reads/writes were causing interleaving of statements ("cannot start a transaction inside a transaction" type errors). This was before adding locking though.